rcl_action/src/rcl_action/graph.c

```diff
@@ -0,0 +1,216 @@
+
+  rcl_ret_t ret;
+  const size_t num_names = topic_names_and_types->names.size;
+  char ** names = topic_names_and_types->names.data;
+
+  // Count number of actions to determine how much memory to allocate
+  size_t num_actions = 0u;
+  for (size_t i = 0u; i < num_names; ++i) {
+    const char * identifier_index = strstr(names[i], action_name_identifier);
+    if (identifier_index) {
+      ++num_actions;
+    }
+  }
+
+  if (0u == num_actions) {
+    return RCL_RET_OK;
+  }
+
+  ret = rcl_names_and_types_init(action_names_and_types, num_actions, allocator);
+  if (RCL_RET_OK != ret) {
+    return ret;
+  }
+
+  ret = RCL_RET_OK;
+
+  // Prune names/types that are not actions (ie. do not contain the suffix)
+  const size_t suffix_len = strlen(action_name_identifier);
+  size_t j = 0u;
+  for (size_t i = 0u; i < num_names; ++i) {
+    const char * identifier_index = strstr(names[i], action_name_identifier);
+    if (identifier_index) {
+      const size_t action_name_len = strlen(names[i]) - suffix_len;
+      char * action_name = rcutils_strndup(names[i], action_name_len, *allocator);
+      if (!action_name) {
+        RCL_SET_ERROR_MSG("Failed to allocate memory for action name");
+        ret = RCL_RET_BAD_ALLOC;
+        break;
+      }
+
+      action_names_and_types->names.data[j] = action_name;
+
+      // Allocate storage for type list
+      rcutils_ret_t rcutils_ret = rcutils_string_array_init(
+        &action_names_and_types->types[j],
+        topic_names_and_types->types[i].size,
+        allocator);
+      if (RCUTILS_RET_OK != rcutils_ret) {
+        RCL_SET_ERROR_MSG(rcutils_get_error_string().str);
+        ret = RCL_RET_BAD_ALLOC;
+        break;
+      }
+
+      // Populate types list
+      for (size_t k = 0u; k < topic_names_and_types->types[i].size; ++k) {
+        char * type_name = topic_names_and_types->types[i].data[k];
+        size_t action_type_len = strlen(type_name);
+        // Trim type name suffix
+        const size_t type_suffix_len = strlen(action_type_identifier);
+        const char * type_identifier_index = strstr(type_name, action_type_identifier);
+        if (type_identifier_index) {
+          action_type_len = strlen(type_name) - type_suffix_len;
+        }
+        // Copy name to output struct
+        char * action_type_name = rcutils_strndup(type_name, action_type_len, *allocator);
+        if (!action_type_name) {
+          RCL_SET_ERROR_MSG("Failed to allocate memory for action type");
+          ret = RCL_RET_BAD_ALLOC;
+          break;
+        }
+        action_names_and_types->types[j].data[k] = action_type_name;
+      }
+      ++j;
+    }
+  }
+
+  // Cleanup if there is an error
+  if (RCL_RET_OK != ret) {
+    rcl_ret_t fini_ret = rcl_names_and_types_fini(action_names_and_types);
+    (void)fini_ret;  // Error already set
+  }
+
+  return ret;
+}
+
+rcl_ret_t
+rcl_action_get_client_names_and_types_by_node(
+  const rcl_node_t * node,
+  rcl_allocator_t * allocator,
```

Comment:
I don't see anything checking if the allocator is valid.

may need to use `RCL_CHECK_ALLOCATOR`

https://github.com/ros2/rcl/blob/d810f12e94ae9498f34ebcf77951b544d21aa10f/rcl/include/rcl/allocator.h#L31-L32

and add a test with a zero initialized allocator

https://github.com/ros2/rcutils/blob/1cb9d67ea7e3ca848d7477e381257776ac00ba85/include/rcutils/allocator.h#L83-L84

Reply:
Added tests 3adab6b